Sources/CoreMetrics/Metrics.swift

```diff
@@ -12,220 +12,230 @@
+    func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> RecorderHandler
+    func makeTimer(label: String, dimensions: [(String, String)]) -> TimerHandler
+}
 
-    private init(_ wrapped: MetricsHandler) {
-        self.wrapped = wrapped
-    }
+// This is the metrics system itself, it's mostly used set the type of the `MetricsHandler` implementation.
+public enum MetricsSystem {
+    fileprivate static let lock = ReadWriteLock()
+    fileprivate static var _handler: MetricsHandler = NOOPMetricsHandler.instance
+    fileprivate static var initialized = false
 
-    public func makeCounter(label: String, dimensions: [(String, String)]) -> Counter {
-        return self.counters.getOrSet(label: label, dimensions: dimensions, maker: self.wrapped.makeCounter)
+    // Configures which `LogHandler` to use in the application.
+    public static func bootstrap(_ handler: MetricsHandler) {
+        self.lock.withWriterLock {
+            precondition(!self.initialized, "metrics system can only be initialized once per process. currently used factory: \(self._handler)")
+            self._handler = handler
+            self.initialized = true
+        }
     }
 
-    public func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> Recorder {
-        let maker = { (label: String, dimensions: [(String, String)]) -> Recorder in
-            self.wrapped.makeRecorder(label: label, dimensions: dimensions, aggregate: aggregate)
-        }
-        return self.recorders.getOrSet(label: label, dimensions: dimensions, maker: maker)
-    }
-
-    public func makeTimer(label: String, dimensions: [(String, String)]) -> Timer {
-        return self.timers.getOrSet(label: label, dimensions: dimensions, maker: self.wrapped.makeTimer)
-    }
-
-    private class Cache<T> {
-        private var items = [String: T]()
-        // using a mutex is never ideal, we will need to explore optimization options
-        // once we see how real life workloads behaves
-        // for example, for short opetations like hashmap lookup mutexes are worst than r/w locks in 99% reads, but better than them in mixed r/w mode
-        private let lock = Lock()
-
-        func getOrSet(label: String, dimensions: [(String, String)], maker: (String, [(String, String)]) -> T) -> T {
-            let key = self.fqn(label: label, dimensions: dimensions)
-            return self.lock.withLock {
-                if let item = items[key] {
-                    return item
-                } else {
-                    let item = maker(label, dimensions)
-                    items[key] = item
-                    return item
-                }
-            }
+    // for our testing we want to allow multiple bootstraping
+    internal static func bootstrapInternal(_ handler: MetricsHandler) {
+        self.lock.withWriterLock {
+            self._handler = handler
```

Comment:
oh ok, that's good; was a bit worried about testing scenarios; this unlocks things... should be fine then, thanks